src/shm/pool/multi.rs

```diff
@@ -0,0 +1,159 @@
+        format: wl_shm::Format,
+        udata: U,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) -> Option<(usize, wl_buffer::WlBuffer, &mut [u8])>
+    where
+        D: Dispatch<wl_buffer::WlBuffer, UserData = U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let mut found_identifier = false;
+        let mut offset = 0;
+        let size = stride * height;
+        let mut index = None;
+        for (i, buffer) in self.buffer_list.iter_mut().enumerate() {
+            if buffer.identifier.eq(identifier) {
+                found_identifier = true;
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.size = buffer.size.max(size as usize);
+                    index = Some(i);
+                }
+            } else if offset > buffer.offset {
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.offset = offset;
+                } else {
+                    index = None;
+                }
+            } else if found_identifier {
+                break;
+            }
+            offset += buffer.size;
+        }
+
+        if found_identifier {
+            offset = self.buffer_list[index?].offset;
+        } else if let Some(b) = self.buffer_list.last() {
+            offset += b.size;
+        }
+
+        if offset + size as usize > self.inner.len
+        && self.resize(offset + 2 * size as usize, conn).is_err() {
+            return None
+        }
+
+        let buffer = self.inner.create_buffer(
+            offset as i32,
+            width,
+            height,
+            stride,
+            format, udata, conn,
+            qh
+        ).ok()?;
+
+        if found_identifier {
+            self.buffer_list[index?].buffer = buffer.clone();
+        } else if index.is_none() {
+            index = Some(self.buffer_list.len());
+            let buffer = Buffer {
+                offset,
+                free: AtomicBool::new(true),
+                buffer: buffer.clone(),
+                size: size as usize,
+                identifier: identifier.clone()
+            };
+            self.buffer_list.push(buffer);
+        }
+
+        let slice = &mut self.inner.mmap()[offset..][..size as usize];
+
+        self.buffer_list[index?].free.swap(false, Ordering::Relaxed);
+
+        Some((offset, buffer, slice))
+    }
+}
+
+impl<I: PartialEq + Clone> DelegateDispatchBase<wl_buffer::WlBuffer> for MultiPool<I> {
+    type UserData = ();
+}
+
+impl<D, I: PartialEq + Clone> DelegateDispatch<wl_buffer::WlBuffer, D> for MultiPool<I>
+where
+    D: Dispatch<wl_buffer::WlBuffer, UserData = Self::UserData>,
+    D: AsMut<MultiPool<I>>
+{
+    fn event(
+        data: &mut D,
+        buffer: &wl_buffer::WlBuffer,
+        event: wl_buffer::Event,
+        _: &Self::UserData,
+        conn: &mut ConnectionHandle,
+        _: &QueueHandle<D>
+    ) {
+        if let wl_buffer::Event::Release = event {
+            buffer.destroy(conn);
```

Comment:
This destroys the object on the server, meaning reuse is impossible.

Reply:
Well if the user is unable to reuse the buffer we gain no optimization here.

Reply:
The underlying buffer is the same it just has a different handle.

Reply:
Iterating over the slice should be fine since the compositor can't create a wl_buffer with the same id if another is still in use.